### PR TITLE
Lua global include

### LINKED ISSUE #11676: Feature request: Global include of LUA scripts
### Description

Please add a feature to authoritative server that allows a common LUA script to be loaded when the server starts.

### Primary use case

Creating new LUA functions that work just like the built-in functions.

### Example

Switch RR value on a given time for a planned migration scenario.

**Solution A so far**

Zone:
```
migratethis     LUA A "os.time() < 1654174980 and '127.0.0.1' or '127.0.1.1'"
migratethat     LUA A "os.time() < 1654174980 and '127.0.0.2' or '127.0.1.2'"
migratewhatever LUA A "os.time() < 1654174980 and '127.0.0.3' or '127.0.1.3'"
```

**Solution B so far**

Zone:
```
switch LUA LUA (" function switchontime(time, before, after) "
                "   if(os.time() < time) then                "
                "     return before                          "
                "   else                                     "
                "     return after                           "
                "   end                                      "
                " end                                        ")

migratethis     LUA A ";include('switch') return switchontime(1654174980, '127.0.0.1', '127.0.1.1')"
migratethat     LUA A ";include('switch') return switchontime(1654174980, '127.0.0.2', '127.0.1.2')"
migratewhatever LUA A ";include('switch') return switchontime(1654174980, '127.0.0.3', '127.0.1.3')"
```

**Proposed solution with new feature**

/etc/pdns/pdns.conf:
```
lua-include-common = /etc/pdns/common-lua.d/*.lua
```
/etc/pdns/common-lua.d/switchontime.lua
```
function switchontime(time, before, after)
  if(os.time() < time) then
    return before
  else
    return after
  end
end
```

Zone:
```
migratethis     LUA A "switchontime(1654174980, '127.0.0.1', '127.0.1.1')"
migratethat     LUA A "switchontime(1654174980, '127.0.0.2', '127.0.1.2')"
migratewhatever LUA A "switchontime(1654174980, '127.0.0.3', '127.0.1.3')"
```

### PR DESCRIPTION
**Pull request by cmouse**
### Short description
Adds ability to load lua scripts globally. Fixes #11676

### Checklist
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by Habbie** [comment]
Makes sense! PR looks roughly okay, the clang-tidy nits appear to make sense. Happy to review after those are covered.

File: pdns/slavecommunicator.cc [hunk 1]
```diff
@@ -699,6 +700,9 @@ void CommunicatorClass::suck(const DNSName &domain, const ComboAddress& remote,
     if(!script.empty()){
       try {
         pdl = make_unique<AuthLua4>();
+        if (!path.empty() && pdl->includePath(path) != 0) {
+          throw PDNSException("Failed to include scripts");
```
Comment (jsoref) [inline pdns/slavecommunicator.cc:704]: Could you include the word `lua`?

File: pdns/auth-main.cc [hunk 2]
```diff
@@ -289,6 +289,7 @@ static void declareArguments()
 
   ::arg().set("lua-prequery-script", "Lua script with prequery handler (DO NOT USE)") = "";
   ::arg().set("lua-dnsupdate-policy-script", "Lua script with DNS update policy handler") = "";
+  ::arg().set("lua-global-include-path", "Path to load Lua scripts for Lua contextes") = "";
```
Comment (jsoref) [inline pdns/auth-main.cc:292]:
```suggestion
  ::arg().set("lua-global-include-path", "Path to load Lua scripts for Lua contexts") = "";
```

File: docs/settings.rst [hunk 3]
```diff
@@ -984,6 +984,17 @@ Not recommended to set this below 3.
...
+.. _setting-lua-global-include-path:
+
+``lua-global-include-path``
+---------------------------
+
+-  String
+-  Default: empty
+-  Example: ``/etc/pdns/lua-global/*.lua``
+
+Path mask to include files to Lua contexts.
```
Comment (Zash) [inline docs/settings.rst:996]: So this loads all the files matching by glob? Every time?

Comment (cmouse) [reply docs/settings.rst:996]: Yes, when the context is created. If the lua context exists, they are not loaded.

Comment (Zash) [reply docs/settings.rst:996]: I think it wouldn't hurt with a couple more words here.

Comment (cmouse) [reply docs/settings.rst]: Better?

File: pdns/lua-base4.cc [hunk 4]
```diff
@@ -36,6 +37,31 @@ void BaseLua4::loadString(const std::string &script) {
   loadStream(iss);
 };
 
+int BaseLua4::includePath(const std::string& path) {
```
Comment (Habbie) [inline pdns/lua-base4.cc:40]: why `int`? the only return statement in here says `0`

File: pdns/recursordist/rec-main.cc [hunk 5]
```diff
@@ -2730,6 +2730,10 @@ static void recursorThread()
       try {
         if (!::arg()["lua-dns-script"].empty()) {
           t_pdl = std::make_shared<RecursorLua4>();
+          const std::string& path = ::arg()["lua-global-include-path"];
```
Comment (Habbie) [inline pdns/recursordist/rec-main.cc:2733]: this is, if I'm counting right, the fourth time I see this line. How about making it an argument to baselua and the derived classes?

**Review by Habbie — CHANGES_REQUESTED** [review]

File: pdns/recursordist/rec-main.cc (hunk 5 above)
Comment (cmouse) [reply pdns/recursordist/rec-main.cc:2733]: you mean accessing the ::arg() from BaseLua?

File: pdns/lua-base4.cc (hunk 4 above)
Comment (cmouse) [reply pdns/lua-base4.cc:40]: meh, i suppose this should fail if loading any of the files fails.

File: pdns/recursordist/rec-main.cc (hunk 5 above)
Comment (Habbie) [reply pdns/recursordist/rec-main.cc:2733]: I was thinking "pass a global include path"

**Comment by Habbie** [comment]
need to check this:

`14:57Z <ottom> Habbie: i seem to remember std::filesystem is not generally available on systems we target`

File: pdns/recursordist/rec-main.cc (hunk 5 above)
Comment (cmouse) [reply pdns/recursordist/rec-main.cc]: maybe i now understood this correctly.

**Comment by cmouse** [comment]
> need to check this:
> 
> `14:57Z <ottom> Habbie: i seem to remember std::filesystem is not generally available on systems we target`
> 
> https://github.com/Habbie/pdns/actions/runs/6852275440

I added check for boost filesystem

**Comment by cmouse** [comment]
@habbie i switched to visit_directory instead of filesystem

File: docs/settings.rst [hunk 6]
```diff
@@ -1046,6 +1046,17 @@ Amount of time (in seconds) between subsequent cleanup routines for pre-computed
...
+.. _setting-lua-global-include-dir:
+
+``lua-global-include-dir``
+---------------------------
+
+-  String
+-  Default: empty
+-  Example: ``/etc/pdns/lua-global/*.lua``
+
+When creating a Lua context, all matching files are loaded and ran in the Lua context.
```
Comment (jsoref) [inline docs/settings.rst:1058]:
```suggestion
When creating a Lua context, all matching files are loaded and run in the Lua context.
```
I think

File: pdns/auth-main.cc [hunk 7]
```diff
@@ -291,6 +291,7 @@ static void declareArguments()
 
   ::arg().set("lua-prequery-script", "Lua script with prequery handler (DO NOT USE)") = "";
   ::arg().set("lua-dnsupdate-policy-script", "Lua script with DNS update policy handler") = "";
+  ::arg().set("lua-global-include-dir", "Path to load Lua scripts for Lua contexts") = "";
```
Comment (jsoref) [inline pdns/auth-main.cc:294]: I'm not sure about this text. Is it only certain file extensions? 

```suggestion
  ::arg().set("lua-global-include-dir", "Path for Lua scripts to include in Lua contexts") = "";
```

File: pdns/recursordist/settings/table.py [hunk 8]
```diff
@@ -1352,6 +1352,16 @@
...
+ ''',
+    },
+    {
+        'name' : 'lua_global_include_path',
+        'section' : 'recursor',
+        'type' : LType.String,
+        'default' : '',
+        'help' : 'More powerful configuration options',
+        'doc' : '''
+ When creating a Lua context, all matching files are loaded and ran in the Lua context.
```
Comment (jsoref) [inline pdns/recursordist/settings/table.py:1364]:
```suggestion
 When creating a Lua context, all matching files are loaded and run in the Lua context.
```

Comment (cmouse) [reply pdns/recursordist/settings/table.py:1364]: ah, right, yes. this needs fixing.

File: docs/settings.rst [hunk 9]
```diff
@@ -1046,6 +1046,18 @@ Amount of time (in seconds) between subsequent cleanup routines for pre-computed
...
+
+``lua-global-include-dir``
+---------------------------
+
+-  String
+-  Default: empty
+-  Example: ``/etc/pdns/lua-global/``
+
+When creating a Lua context, scan this directory for additional lua files. All files that end with
+.lua are loaded in order using ``POSIX`` as locale with Lua scripts.
```
Comment (Habbie) [inline docs/settings.rst:1059]: this text needs some improving, will ponder

File: pdns/lua-base4.cc [hunk 10]
```diff
@@ -15,24 +17,60 @@
...
 void BaseLua4::loadString(const std::string &script) {
   std::istringstream iss(script);
-  loadStream(iss);
+  loadStream(iss, true);
+};
+
+void BaseLua4::includePath(const std::string& directory) {
+  std::vector<std::string> vec;
+  const std::string& suffix = "lua";
+	  auto directoryError = pdns::visit_directory(directory, [this, &directory, &suffix, &vec]([[maybe_unused]] ino_t inodeNumber, const std::string_view& name) {
```
Comment (Habbie) [inline pdns/lua-base4.cc:40]: indent seems weird

File: pdns/lua-base4.cc [hunk 11]
```diff
@@ -289,10 +327,11 @@ void BaseLua4::prepareContext() {
   d_lw->writeVariable("pdns", d_pd);
 }
 
-void BaseLua4::loadStream(std::istream &is) {
+void BaseLua4::loadStream(std::istream &is, bool doPostLoad) {
```
Comment (Habbie) [inline pdns/lua-base4.cc:330]: tidy would like a longer var name here

File: pdns/packethandler.cc [hunk 12]
```diff
@@ -71,14 +71,15 @@ PacketHandler::PacketHandler():B(g_programname), d_dk(&B)
...
   if(fname.empty())
   {
     d_pdl = nullptr;
   }
   else
   {
-    d_pdl = std::make_unique<AuthLua4>();
-    d_pdl->loadFile(fname); // XXX exception handling?
+    d_pdl = std::make_unique<AuthLua4>(::arg()["lua-global-include-dir"]);
+    d_pdl->loadFile(fname);
```
Comment (Habbie) [inline pdns/packethandler.cc:82]: why is the comment gone?

**Review by Habbie — COMMENTED** [review]
this seems sensible. Want to handle clang-tidy's complaints? Then I'll do a fuller review.

File: pdns/packethandler.cc (hunk 12 above)
Comment (cmouse) [reply pdns/packethandler.cc:82]: turns out the exception handling is just fine.

File: pdns/lua-base4.cc (hunk 11 above)
Comment (cmouse) [reply pdns/lua-base4.cc:330]: it feels like can of worms, but i'll give it a try.

File: docs/settings.rst (hunk 9 above)
Comment (cmouse) [reply docs/settings.rst:1059]: I copied this from include-dir

Comment (Habbie) [reply docs/settings.rst:1059]: I see. Then both need improvement. Not your problem then :)

**Review by Habbie — APPROVED** [review]

**Pull request merged**